crypto/envelope.go

```diff
@@ -0,0 +1,191 @@
+package crypto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"github.com/libp2p/go-buffer-pool"
+	pb "github.com/libp2p/go-libp2p-core/crypto/pb"
+)
+
+// SignedEnvelope contains an arbitrary []byte payload, signed by a libp2p peer.
+// Envelopes are signed in the context of a particular "domain", which is a string
+// specified when creating and verifying the envelope. You must know the domain
+// string used to produce the envelope in order to verify the signature and
+// access the payload.
+type SignedEnvelope struct {
+
+	// The public key that can be used to verify the signature and derive the peer id of the signer.
+	publicKey PubKey
+
+	// A binary identifier that indicates what kind of data is contained in the payload.
+	// TODO(yusef): enforce multicodec prefix
+	payloadType []byte
+
+	// The envelope payload.
+	payload []byte
+
+	// The signature of the domain string, type hint, and payload.
+	signature []byte
```

Comment:
Why not make the fields public? This way we can avoid the accessor functions and use it as a proper struct.